[Bug]: Copilot adaptor not working after 30 minutes

### Did you check the docs and existing issues?

- [x] I have read all the docs
- [x] I have updated the plugin to the latest version before submitting this issue
- [x] I have searched for existing issues/discussions
- [x] I have searched the existing issues of plugins related to this issue

### Neovim version (nvim -v)

NVIM v0.10.2 Release

### Operating system/version

MacOs 15.0

### Describe the bug

Copilot adapter does not refresh it's expiry time of the `_github_token`. After the first expiry, the token will not refresh. 
Restarting the nvim instance will solve this for the next 30 minutes.

### Steps To Reproduce

1. Open chat buffer and ask a quesiton
2. wait 30 min for github token to expire
3. Send another request to copilot and it will fail with:
```
[ERROR] 2025-03-01 22:05:53
Error: unauthorized: token expired
```

### Expected Behavior

I would expect the chat request to not fail with `Error: unauthorized: token expired`

### Log output

```
[DEBUG] 2025-03-01 21:31:32
Variable found: buffer
[INFO] 2025-03-01 21:31:32
Chat request started
[INFO] 2025-03-01 21:31:32
ELAPSED TIME: 35.657625
[DEBUG] 2025-03-01 21:31:32
Authorizing GitHub Copilot token
[INFO] 2025-03-01 21:31:33
Request body file: /var/folders/zv/g_ls02w15dj90l9503dmqs940000gn/T/nvim.minescoviciu/OCG9xK/0.json
[DEBUG] 2025-03-01 21:31:33
Request:
{ "-sSL", "-D", "/tmp/plenary_curl_f4574a9c.headers", "-X", "POST", "-H", "Authorization: Bearer tid=33f659eaaa3e02209ebf80b89aa9f706;exp=1740859293;sku=yearly_subscriber;proxy-ep=proxy.individual.githubcopilot.com;st=dotcom;chat=1;cit=1;malfil=1;editor_preview_features=1;8kp=1;ip=188.24.173.153;asn=AS8708:73fa0e5c13665cfeba96b5e34212f21e30be7231a9b5b61142dac6b1627d1c23", "-H", "Content-Type: application/json", "-H", "Copilot-Integration-Id: vscode-chat", "-H", "Editor-Version: Neovim/0.10.2", "-d", "@/var/folders/zv/g_ls02w15dj90l9503dmqs940000gn/T/nvim.minescoviciu/OCG9xK/0.json", "--retry", "3", "--retry-delay", "1", "--keepalive-time", "60", "--connect-timeout", "10", "--tcp-nodelay", "--no-buffer", "https://api.githubcopilot.com/chat/completions" }
[INFO] 2025-03-01 21:31:37
Chat request finished
[INFO] 2025-03-01 22:05:52
Chat request started
[INFO] 2025-03-01 22:05:52
ELAPSED TIME: 33.998459
[DEBUG] 2025-03-01 22:05:52
Authorizing GitHub Copilot token
[INFO] 2025-03-01 22:05:52
Request body file: /var/folders/zv/g_ls02w15dj90l9503dmqs940000gn/T/nvim.minescoviciu/OCG9xK/1.json
[DEBUG] 2025-03-01 22:05:52
Request:
{ "-sSL", "-D", "/tmp/plenary_curl_b7fb4d7b.headers", "-X", "POST", "-H", "Authorization: Bearer tid=33f659eaaa3e02209ebf80b89aa9f706;exp=1740859293;sku=yearly_subscriber;proxy-ep=proxy.individual.githubcopilot.com;st=dotcom;chat=1;cit=1;malfil=1;editor_preview_features=1;8kp=1;ip=188.24.173.153;asn=AS8708:73fa0e5c13665cfeba96b5e34212f21e30be7231a9b5b61142dac6b1627d1c23", "-H", "Content-Type: application/json", "-H", "Copilot-Integration-Id: vscode-chat", "-H", "Editor-Version: Neovim/0.10.2", "-d", "@/var/folders/zv/g_ls02w15dj90l9503dmqs940000gn/T/nvim.minescoviciu/OCG9xK/1.json", "--retry", "3", "--retry-delay", "1", "--keepalive-time", "60", "--connect-timeout", "10", "--tcp-nodelay", "--no-buffer", "https://api.githubcopilot.com/chat/completions" }
```

### File to Reproduce

```lua
---@diagnostic disable: missing-fields

--[[
NOTE: Set the config path to enable the copilot adapter to work.
It will search the following paths for a token:
  - "$CODECOMPANION_TOKEN_PATH/github-copilot/hosts.json"
  - "$CODECOMPANION_TOKEN_PATH/github-copilot/apps.json"
--]]
vim.env["CODECOMPANION_TOKEN_PATH"] = vim.fn.expand("~/.config")

vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

-- Your CodeCompanion setup
local plugins = {
  {
    "olimorris/codecompanion.nvim",
    dependencies = {
      { "nvim-treesitter/nvim-treesitter", build = ":TSUpdate" },
      { "nvim-lua/plenary.nvim" },
      -- Test with blink.cmp
      {
        "saghen/blink.cmp",
        lazy = false,
        version = "*",
        opts = {
          keymap = {
            preset = "enter",
            ["<S-Tab>"] = { "select_prev", "fallback" },
            ["<Tab>"] = { "select_next", "fallback" },
          },
          cmdline = { sources = { "cmdline" } },
          sources = {
            default = { "lsp", "path", "buffer", "codecompanion" },
          },
        },
      },
      -- Test with nvim-cmp
      -- { "hrsh7th/nvim-cmp" },
    },
    opts = {
      --Refer to: https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/config.lua
      strategies = {
        --NOTE: Change the adapter as required
        chat = { adapter = "copilot" },
        inline = { adapter = "copilot" },
      },
      opts = {
        log_level = "DEBUG",
      },
    },
  },
}

require("lazy.minit").repro({ spec = plugins })

-- Setup Tree-sitter
local ts_status, treesitter = pcall(require, "nvim-treesitter.configs")
if ts_status then
  treesitter.setup({
    ensure_installed = { "lua", "markdown", "markdown_inline", "yaml" },
    highlight = { enable = true },
  })
end

-- Setup nvim-cmp
-- local cmp_status, cmp = pcall(require, "cmp")
-- if cmp_status then
--   cmp.setup({
--     mapping = cmp.mapping.preset.insert({
--       ["<C-b>"] = cmp.mapping.scroll_docs(-4),
--       ["<C-f>"] = cmp.mapping.scroll_docs(4),
--       ["<C-Space>"] = cmp.mapping.complete(),
--       ["<C-e>"] = cmp.mapping.abort(),
--       ["<CR>"] = cmp.mapping.confirm({ select = true }),
--       -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
--     }),
--   })
-- end
```

### Have you provided and tested with a repro.lua file?

- [x] Yes, I have tested and provided a repro.lua file

Yeah, after searching far and wide - I can't seem to find anyone with the same issue as me, this is the message I get:

[ERROR] 2025-07-05 00:27:38                                                                                                                                                              
Error parsing the response from https://api.individual.githubcopilot.com/models.                                                                                                         ╰──────────────╯
Error: unauthorized: token expired

This was after about 40 minutes of working, longest prompting break was around 5 minutes

--- a/lua/codecompanion/adapters/copilot/helpers.lua
+++ b/lua/codecompanion/adapters/copilot/helpers.lua
@@ -30,9 +30,9 @@ end
 ---Get a list of available Copilot models
 ---@param adapter CodeCompanion.Adapter
 ---@param get_and_authorize_token_fn function Function to get and authorize token
----@param github_token table The github token object
+---@param oauth_token string The oauth token
 ---@return table
-function M.get_models(adapter, get_and_authorize_token_fn, github_token)
+function M.get_models(adapter, get_and_authorize_token_fn, oauth_token)
   if _cached_models and _cache_expires and _cache_expires > os.time() then
     return _cached_models
   end
@@ -44,13 +44,16 @@ function M.get_models(adapter, get_and_authorize_token_fn, github_token)
     _cached_adapter = adapter
   end
 
-  get_and_authorize_token_fn(adapter)
-  local url = github_token.endpoints.api or "https://api.githubcopilot.com"
+  if not get_and_authorize_token_fn(adapter) then
+    return {}
+  end
+
+  local url = "https://api.githubcopilot.com/models"
   local headers = vim.deepcopy(_cached_adapter.headers)
-  headers["Authorization"] = "Bearer " .. github_token.token
+  headers["Authorization"] = "Bearer " .. oauth_token
 
   local ok, response = pcall(function()
-    return curl.get(url .. "/models", {
+    return curl.get(url, {
       sync = true,
       headers = headers,
       insecure = config.adapters.opts.allow_insecure,
@@ -58,13 +61,13 @@ function M.get_models(adapter, get_and_authorize_token_fn, github_token)
     })
   end)
   if not ok then
-    log:error("Could not get the Copilot models from " .. url .. "/models.\nError: %s", response)
+    log:error("Could not get the Copilot models from " .. url .. ".\nError: %s", response)
     return {}
   end
 
   local ok, json = pcall(vim.json.decode, response.body)
   if not ok then
-    log:error("Error parsing the response from " .. url .. "/models.\nError: %s", response.body)
+    log:error("Error parsing the response from " .. url .. ".\nError: %s", response.body)
     return {}
   end
 

--- a/lua/codecompanion/adapters/copilot/init.lua
+++ b/lua/codecompanion/adapters/copilot/init.lua
@@ -300,7 +300,7 @@ return {
             return { ["gpt-4.1"] = { opts = {} } } -- fallback
           end
         end
-        return copilot_helper.get_models(self, get_and_authorize_token, _github_token)
+        return copilot_helper.get_models(self, get_and_authorize_token, _oauth_token)
       end,
     },
     ---@type CodeCompanion.Schema

--- a/tests/adapters/copilot/test_helpers.lua
+++ b/tests/adapters/copilot/test_helpers.lua
@@ -120,4 +120,77 @@ T["Copilot Helper Stats"]["show_copilot_stats handles overage_permitted for prem
   h.eq(true, found_overage, "Expected overage_permitted to be displayed in stats")
 end
 
+T["Copilot Helper Get Models"] = new_set()
+
+T["Copilot Helper Get Models"]["retrieves models with correct structure"] = function()
+  local mock_models = {
+    {
+      id = "model1",
+      capabilities = { type = "chat", supports = { streaming = true, tool_calls = true, vision = true } },
+      model_picker_enabled = true,
+    },
+    {
+      id = "model2",
+      capabilities = { type = "chat", supports = {} },
+      model_picker_enabled = true,
+    },
+    {
+      id = "model3",
+      capabilities = { type = "completion" },
+      model_picker_enabled = false,
+    },
+  }
+  local curl = require("plenary.curl")
+  local original_curl_get = curl.get
+  curl.get = function(url, _)
+    if url == "https://api.githubcopilot.com/models" then
+      return {
+        body = vim.json.encode({ data = mock_models }),
+        status = 200,
+      }
+    else
+      return { status = 404, body = "Not Found" }
+    end
+  end
+  local utils = require("codecompanion.utils.adapters")
+  local original_refresh_cache = utils.refresh_cache
+  local refresh_cache_mock = function()
+    return 0
+  end
+  utils.refresh_cache = refresh_cache_mock
+  local mock_adapter = { url = "https://api.githubcopilot.com", headers = {} }
+  local mock_get_and_authorize_token = function()
+    return true -- Simulate successful token retrieval
+  end
+  local mock_oauth_token = "mock_oauth_token"
+
+  local models = copilot_helper.get_models(mock_adapter, mock_get_and_authorize_token, mock_oauth_token)
+
+  h.eq(2, vim.tbl_count(models), "Expected two models to be returned")
+  h.eq({
+    opts = {
+      can_stream = true,
+      can_use_tools = true,
+      has_vision = true,
+    },
+  }, models.model1, "First model should match")
+  h.eq({ opts = {} }, models.model2, "Second model should match")
+
+  -- Restore original function
+  utils.refresh_cache = original_refresh_cache
+  curl.get = original_curl_get
+end
+
+T["Copilot Helper Get Models"]["returns empty table when Copilot token refresh is not OK"] = function()
+  local mock_adapter = { url = "https://api.githubcopilot.com", headers = {} }
+  local mock_get_and_authorize_token = function()
+    return false -- Simulate unsuccessful token retrieval
+  end
+  local mock_oauth_token = "mock_oauth_token"
+
+  local models = copilot_helper.get_models(mock_adapter, mock_get_and_authorize_token, mock_oauth_token)
+
+  h.eq(0, vim.tbl_count(models), "Expected no models to be returned")
+end
+
 return T

